nameserver/zone.go

```diff
@@ -585,30 +583,45 @@ func (zone *ZoneDb) AddRecord(ident string, name string, ip net.IP) (err error)
 	return
 }
 
-// Delete all records for an ident
-func (zone *ZoneDb) DeleteRecordsFor(ident string) (err error) {
-	zone.mx.Lock()
-	defer zone.mx.Unlock()
-	Debug.Printf("[zonedb] Deleting records for ident '%s'", ident)
-	zone.deleteIdent(ident)
-	return
-}
-
-// Delete an IP in a ident
-func (zone *ZoneDb) DeleteRecord(ident string, ip net.IP) (err error) {
+// Delete matching records (uninitialised values act as wildcards)
+func (zone *ZoneDb) DeleteRecords(ident string, name string, ip net.IP) int {
 	zone.mx.Lock()
 	defer zone.mx.Unlock()
 
-	Debug.Printf("[zonedb] Deleting record '%s'/[%s]", ident, ip)
-	if _, found := zone.idents[ident]; found {
-		err = zone.idents[ident].deleteIP(ipToIPv4(ip))
-		if zone.idents[ident].empty() {
-			zone.deleteIdent(ident)
+	count := 0
+
+	if name != "" {
+		name = dns.Fqdn(name)
+	}
+
+	for identK, identV := range zone.idents {
+		if ident == "" || ident == identK {
+			for nameK, nameV := range identV.names {
+				nameModified := false
+				if name == "" || name == nameK {
+					for ipv4K, ipv4V := range nameV.ipv4 {
+						if ip.Equal(net.IP{}) || ip.Equal(ipv4K.toNetIP()) {
+							ipv4V.notifyIPObservers()
+							delete(nameV.ipv4, ipv4K)
+							nameModified = true
+							count++
+						}
+					}
+				}
+				if nameModified {
+					nameV.notifyNameObservers()
+					if nameV.empty() {
+						delete(identV.names, nameK)
+					}
+				}
+			}
+		}
```

Comment:
I considered it, but given you have to iterate over all the IPs anyway when you remove a name to notify observers, I decided against. I've a feeling that it'll be a bit messy because you'll end up pulling the innards of the name and IP loops out into separate functions to avoid duplication between the direct lookup/full scan paths - I'll have a play and see what it looks like.